### PR TITLE
felt cute, have a suffix array

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

no one reads this anyways
might add other languages if i'm feeling up to it

### CODE REVIEW

**Review by ryanchou-dev — DISMISSED** [review]
wowowow kevin carrying usaco.guide 🔥🔥🔥

very cool content, tables look great!

**Review by envyaims — APPROVED** [review]
this is cool

File: content/6_Advanced/Suffix_Array.mdx [hunk 1]
```diff
@@ -35,7 +31,211 @@ frequency: 1
...
+
+The general philosophy behind the construction of a suffix array is to sort incrementally.
+We first start by comparing the first character of each suffix, and then double the amount we compare until we're comparing the entire length of the string.
+If this sounds really abstract, don't worry!
+All the implementation details will be hammered out below.
+
+### Initial Transformation
+
+For convenience, let's start by appending a "null character" of sorts to the end of the string.
+This acts as a tiebreaker of sorts and will ensure we never hit the end of a suffix while comparing two suffixes.
```
Comment (danielzsh) [inline content/6_Advanced/Suffix_Array.mdx:65]: english teacher ahh comment but "of sorts" is repeated twice in close proximity 😅

File: content/6_Advanced/Suffix_Array.mdx [hunk 2]
```diff
@@ -35,7 +31,211 @@ frequency: 1
...
+To get to the point where we efficiently compare suffixes, let's first do some analysis.
+
+Say we just sorted all the suffixes by the first $2$ characters, we're now comparing two suffixes that start at $a$ and $b$ respectively, each of length $4$:
+
+<center>
+
+| Suffix | String |
+| ------ | ------ |
+| a      | `????` |
+| b      | `????` |
```
Comment (danielzsh) [inline content/6_Advanced/Suffix_Array.mdx:119]: Why are both strings `????`? should we provide an example instead?

File: content/6_Advanced/Suffix_Array.mdx [hunk 3]
```diff
@@ -35,7 +31,211 @@ frequency: 1
...
+If two half-baked suffixes are equal, then they would get the same number.
+These numbers are usually called [equivalence classes](https://en.wikipedia.org/wiki/Equivalence_class).
+
+Applying this augmentation to our array sorted by the first character, we would now have:
+
+<center>
+
+| Eq. Class | Suffix | String            |
+| --------- | ------ | ----------------- |
+| 0         | 7      | `$abcbcba`	     |
```
Comment (danielzsh) [inline content/6_Advanced/Suffix_Array.mdx:139]: umm what happened to the `|` on the right

File: content/6_Advanced/Suffix_Array.mdx [hunk 4]
```diff
@@ -35,7 +31,211 @@ frequency: 1
...
+| a      | `????` |
+| b      | `????` |
+
+</center>
+
+If the first $2$ characters of `a` are already less than or greater than those of `b`, then nothing changes.
+However, if the two are equal, we're going to have to compare the right half.
+
+How might we do this?
+Well, the cool thing is that since we just sorted *all* suffixes by the first $2$, we actually have information about the right halves already!
```
Comment (danielzsh) [inline content/6_Advanced/Suffix_Array.mdx:127]: do we? doesn't this give us info about the left halves?

**Review by danielzsh — CHANGES_REQUESTED** [review]

**Review by ryanchou-dev — APPROVED** [review]

**Review by danielzsh — APPROVED** [review]
we up 🔥

**Pull request merged**